### PR TITLE
Developmental

### PR DESCRIPTION
**Pull request by magoed12**

### CODE REVIEW

**Review by scott-bree — APPROVED** [review]
Good development of code can be seen, there seems to be room to be more consistent with implicitly typed and explicitly typed variables. Good use of comments to describe code which helps the code be more readable. The code meets the brief requirements. Error checking can be seen in the test class.

File: CMP1903_A1_2324/Game.cs [hunk 1]
```diff
@@ -1,21 +1,28 @@
...
+            die1.Roll();
+            Thread.Sleep(1);// the sleep function is used because if the random object that is declared is used too soon after each other it will produce the same number for all die
+            Die die2 = new Die();
+            die2.Roll();
+            Thread.Sleep(1);
+            Die die3 = new Die();
+            die3.Roll();
+            Thread.Sleep(1);
+            return (die1: die1.Dieval , die2: die2.Dieval, die3: die3.Dieval, sum: die1.Dieval + die2.Dieval + die3.Dieval);// retruns all values in the form (Die1,Die2,Die3,sum of all die)
+        }
```
Comment (JoshyTM123) [inline CMP1903_A1_2324/Game.cs:26]: Good commenting, code makes sense and follows the c# style sheet

File: CMP1903_A1_2324/Die.cs [hunk 2]
```diff
@@ -8,16 +8,19 @@ namespace CMP1903_A1_2324
...
+            get { return dieval; }
+            set { dieval = value; }
+        }
+        public int Roll()// roll method, chooses random number between 1 and 6 then returns it
+        {
+            Random random = new Random();// creates a object called random that is thern used to generate a random number between 1 - 6
+            dieval = random.Next(1, 7);
+            Console.WriteLine(dieval);
+            return dieval;// returns an integer from the method
+        }
```
Comment (JoshyTM123) [inline CMP1903_A1_2324/Die.cs:24]: Commenting was done nicely and code follows c# style structure

File: CMP1903_A1_2324/Testing.cs [hunk 3]
```diff
@@ -1,20 +1,23 @@
...
+        public void Test()// testing method called Test called in the main program
+        {
+            Game game = new Game();// creates a Game object simply called game
+            var values = game.RunGame();// as the RunGame returns as a a tuple, this lines saves the tuple to a local variable
+            Console.WriteLine(values.Item4);// prints the sum of the numbers
+            Debug.Assert(values.Item4 < 19 && values.Item4 > 2, "The sum is incorrect");// error handling using debug.assert, checks the sum and the die are possible given that it is generating random numbers between 1-6
+            Debug.Assert(values.Item1 < 7 && values.Item1 > 0, "A die roll is invalid");
+            Debug.Assert(values.Item2 < 7 && values.Item2 > 0, "A die roll is invalid");
+            Debug.Assert(values.Item3 < 7 && values.Item3 > 0, "A die roll is invalid");
+        }
```
Comment (JoshyTM123) [inline CMP1903_A1_2324/Testing.cs:21]: Code follows coding style correctly, comments used well

File: CMP1903_A1_2324/Program.cs [hunk 4]
```diff
@@ -8,12 +8,10 @@ namespace CMP1903_A1_2324
...
-        static void Main(string[] args)
+        static void Main(string[] args)// main runs when program is run
         {
-            /*
-             * Create a Game object and call its methods.
-             * Create a Testing object to verify the output and operation of the other classes.
-             */
+            Testing testing = new Testing();// creates a testing object
+            testing.Test();// calls the Test method within the newly created testing object
         }
```
Comment (JoshyTM123) [inline CMP1903_A1_2324/Program.cs:15]: You are missing the 'create a game object and call its methods' part out here!

**Review by JoshyTM123 — CHANGES_REQUESTED** [review]
In the program class, you have not created a game object and called it's methods. This should probably be added, but other than that you've used good commenting, your code follows the c# coding style, your code is fairly easy to read, and in the test class you've used a good amount of error handling.

**Pull request merged**